### ISSUE TITLE
Consider explicitly prohibiting exception propagation by one-way execution functions

### ISSUE THREAD

**Issue by chriskohlhoff** [executors/issaquah_2016#189]
One way executors and exceptions
================================

What should happen when a submitted function throws an exception? Currently, the wording explicitly notes that the behaviour is specific to an executor (a.k.a. deliberately left unspecified by the executor requirements). This means that the exception is allowed to be propagated via the execution function (e.g. execute()). For example, given a trivial inline executor:

    class my_inline_executor
    {
    public:
      // ...

      template <class Function>
      void execute(Function f) const
      {
        f();
      }
    };

which may be used as follows:

    ex.execute([]{ throw 42; });

the caller must expect that the exception may be thrown to the caller of `execute()`. This inline executor conforms to the current requirements.

This is the existing practice of execution functions in Boost.Asio and the Networking TS. The rationale is that, in event-driven programs, an exception that escapes from an event handler is a potentially fatal condition. It should be handled as soon as possible in order to initiate graceful shutdown or recovery. Since event handlers often execute other handlers:

    void outer_event_handler()
    {
      // ...
      auto ex = ...;
      ex.execute(inner_event_handler);
      // ...
    }

an exception that escapes from the inner handler should be allowed to unwind the outer handler, to return control to the outer event loop as soon as possible.

The proposal is that we should *explicitly prohibit* the propagation of exceptions through the one-way execution functions. (In practice, this change applies only to the `execute()` funtion, as the `post()` and `defer()` execution functions will not invoke the function inline.)

*Note:* This does not mean that one-way execution functions become noexcept. Execution functions may still throw exceptions related to the submission of a function to an executor, e.g. failure to allocate memory to hold the type-erased function, failure to acquire a mutex, etc.

Impact on executor and execution context design
-----------------------------------------------

Conceptually, an execution context defines the default exception handling "back channel". The exact mechanism of this channel is specific to the concrete execution context. Some possible back channel designs are:

* An execution context allows users to register an exception handler callback.

* An execution context provides a function to collect any queued exceptions.

* A Networking TS `io_context` (re-)throws the exceptions out of any calls to `run()` (or the any other run functions).

* A static thread pool may simply "channel" the exceptions straight to `std::terminate()`.

The benefit of the proposed change is that it ensures that, by default, *all* exceptions will go to the channel. It eliminates the potential inconsistency when `execute()` invokes a function inline. (Note that an executor adapter is still permitted to override the exception handling back channel; see below.)

Effect of proposed change on generic code
-----------------------------------------

Consider again the chained event handler example from above:

    void outer_event_handler()
    {
      // ... 1
      auto ex = ...; // 2
      ex.execute(inner_event_handler); // 3
      // ... 4
    }

This code is generic with respect to the type of the executor `ex`, so it does not know whether the `execute()` function will invoke the `inner_event_handler` inline. That is, even if `inner_event_handler` always throws, the executor may simply post the function for later execution. Therefore, even under the current model, the code cannot assume that an exception from `inner_event_handler` will prevent the line marked 4 from being reached.

Conversely, the execution function on line #3 can still throw for other reasons. Thus the author of `outer_event_handler` must still assume that line 4 may not be reached.

For these reasons, I do not think the proposed change affects the way these generic functions are written.

However, if the event handler was written to catch exceptions:

    void outer_event_handler()
    {
      try
      {
        // ...
        auto ex = ...;
        ex.execute(inner_event_handler);
        // ...
      }
      catch (...)
      {
        // ...
      }
    }

the proposed change *will* prevent an `inner_event_handler`-originated exception from being swallowed by `outer_event_handler`. This may be a benefit: if the above code was only ever tested with non-blocking `execute()` implementations, it won't start silently swallowing them when switched to an inline executor.

Effect of proposed change on inline executors
---------------------------------------------

With this change, implementers of trivial executors must now take additional care to ensure that exceptions do not escape. The most likely approach will be to call `std::terminate`:

    class my_inline_executor
    {
    public:
      // ...

      template <class Function>
      void execute(Function f) const
      {
        try
        {
          f();
        }
        catch (...)
        {
          std::terminate();
        }
      }
    };

or perhaps simply:

    class my_inline_executor
    {
    public:
      // ...

      template <class Function>
      void execute(Function f) const noexcept
      {
        f();
      }
    };

Effect of proposed change on adapters
-------------------------------------

Most adapters should be unaffected by the proposed change. Consider an adapter that decorates function invocation with logging:

    template <class Ex>
    class my_logging_executor
    {
      Ex ex_;

    public:
      explicit my_logging_executor(Ex ex) : ex_(ex) {}

      // ...

      template <class Function>
      void execute(Function f) const
      {
        ex_.execute(
            [f=std::move(f)]() mutable
            {
              std::cout << "before\n";
              f();
              std::cout << "after\n";
            });
      };
    };

If the submitted function `f` throws an exception, it will simply be allowed to escape from the lamdba. Responsibility for the exception remains with the adapted executor `ex_`, which will pass it to whatever "back channel" it specifies.

Adapters are still permitted (as they are under the existing behaviour) to override exception handling with their own back channel. For example:

    template <class Ex>
    class my_exception_handling_executor
    {
      Ex ex_;
      std::function<void(std::exception_ptr)> eh_;

    public:
      my_exception_handling_executor(Ex ex, std::function<void(std::exception_ptr)> eh) : ex_(ex), eh_(eh) {}

      // ...

      template <class Function>
      void execute(Function f) const
      {
        ex_.execute(
            [f=std::move(f), eh_]() mutable
            {
              try
              {
                f();
              }
              catch (...)
              {
                eh_(std::current_exception());
              }
            });
      };
    };

Performance impact
------------------

The performance overhead of this change on a trivial inline executor can be measured using the following benchmark:

    #include <exception>

    class my_inline_executor
    {
    public:
      template <class F>
        void execute(F f) const
        {
    #ifdef PROPOSED
          try
    #endif
          {
            f();
          }
    #ifdef PROPOSED
          catch (...)
          {
            std::terminate();
          }
    #endif
        }
    };

    void my_event_handler(); // an empty function defined in another translation unit

    int main()
    {
      my_inline_executor ex;
      for (int i = 0; i < 1000000000; ++i)
        ex.execute(my_event_handler);
    }

Using:

* g++ version 6.3.0
* Linux 4.9.0-3-amd64
* Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz

there was no difference in total program execution time (1.62s).

To measure the likely impact on event driven programs, the Boost.Asio library was modified (conditionally, using `#ifdef PROPOSED` as above) as follows:

* A `std::exception_ptr` object was added to the library's existing thread-local run state, to store any queued exception.
* A try/catch block was added to the `io_service::dispatch()` function (the direct equivalent of `execute()` in the executor proposal), to catch any exception produced by inline execution of the submitted function. If one is caught, the exception is added to the thread-local queue.
* The `io_service::run()` function was modified to check whether the queue is non-empty after every function execution, and if so to rethrow the queued exception.

And the following benchmark was run:

    #include <boost/asio.hpp>

    boost::asio::io_service io_service(1);
    int count = 0;

    void outer_handler();

    void inner_handler()
    {
      io_service.post(outer_handler);
    }

    void outer_handler()
    {
      if (count++ < 1000000000)
        io_service.dispatch(inner_handler);
    }

    int main()
    {
      io_service.post(outer_handler);
      io_service.run();
    }

Without the proposed change, run time was 31.6s. With the change, run time was 32.1s. This represents a cost of approximately 0.5ns per iteration.

**Comment by jaredhoberock** [comment]
Thanks Chris, nice analysis.

One thing that occurred to me is that we may have some difficulty distinguishing between user exceptions (i.e., exceptions escaping the user's submitted function) and submission exceptions (i.e., exceptions arising from the act of work creation, distinct from exceptions created by the work).

The issue I'm thinking about is one-way executor adapters that wrap two-way executors. Consider this example:

    template<class TwoWayExecutor>
    class my_one_way_executor_adapter
    {
      public:
        template<class class F>
        void execute(F f)
        {
          // XXX sync_execute() throws user exceptions and submission exceptions
          //         how to swallow user exceptions while propagating submission exceptions?
          execution::sync_execute(f);
        };

      private:
        TwoWayExecutor ex_;
    };

If `my_one_way_executor_adapter` is obligated to hide user exceptions, it seems difficult to do that generically while not also hiding submission exceptions.

**Comment by AerialMantis** [comment]
Thanks Chris, this sounds reasonable to me.

I think a good approach could be to require that one-way execution functions throw exceptions to a callback object if one is provided and if not call `std::terminate()`.

A couple of questions that come to mind though (and I'm just brain dumping here):
* How should such a callback object be provided, i.e. I think it makes sense to provide it to the execution context, however, some implementations may not provide a context, so it must also be possible to provide it to an executor, or we could require that it always be provided to an executor.
* Some executors may provide both one-way and two-way execution functions and therefore it will be possible for the executor or execution context to have a callback object when calling a two-way execution function. What should happen in this case, should the callback object take precedence for handling exceptions over propagating via the future or should it propagate exceptions via the future as normal? Or should we have some kind of hint for specifying the preferred exception handling method?

I also agree that we need to be able to distinguish between exceptions originating from submission and exceptions originating from the user's function.

The way exceptions are handled in SYCL is by having an `async_handler` (essentially a callback object). In SYCL all tasks are submitted asynchronously so we have this distinction of synchronous and asynchronous exceptions. Synchronous exceptions are those that originate in the submission of a task, these exceptions are propagated to the caller so they can be handled immediately. Asynchronous exceptions are those that originate from the asynchronous task, these exceptions are propagated to the `async_handler` if one is provided when the queue is prompted.

We use `exception_list` here as we adopted it before it was removed from the parallelism TS.

```cpp
// Create a queue with an async_handler which rethrows the exceptions.
queue q([&](exception_list eL){ for (auto& e : eL) { std::rethrow_exception(e); } })

try {
  // Submit asynchronous task
  q.submit([&](handler& cgh) {
    cgh.single_task<class kernel0([=]() {
      /* ... */
    });
  });
  // Wait on task to complete and throw exceptions
  q.wait_and_throw();
}
catch (cl::sycl::exception& e) {
  handle_exception(e);
}
```

SYCL doesn't permit user exceptions as it is based on OpenCL where exceptions are not allowed within a kernel but it does handle exceptions in en-queuing the task asynchronously. I think the way asynchronous exceptions are handled here could be relevant here.

Perhaps this could be resolved by requiring that exceptions originating from submission are propagated to the caller but exceptions originating from the user's function are propagated to the callback object. And by requiring that any execution function, including an adaption, is responsible for handling exceptions appropriately for that execution function.

So in the case that Jared described above the `execute()`, as it's a one-way execution function it would be required to propagate any user exceptions from `sync_execute` to the callback object and propagate any exceptions originating from submission directly to the caller. Though this would require a guarantee that any exceptions originating from submission be thrown inline and not asynchronously from another thread.

```cpp
template<class TwoWayExecutor>
class my_one_way_executor_adapter
{
  public:
    template<class class F>
    void execute(F f)
    {
      try {
        auto fut = execution::sync_execute(ex_, f);
        // propagate user exceptions from sync_execute from the future to the callback object.
        this->throw_to_callback(fut);
      }
      catch (...)
      {
        // propagate submission exceptions to the caller
        std::rethrow_exception(std::current_exception())
      }
    };

  private:
    TwoWayExecutor ex_;
};
```

**Comment by jaredhoberock** [comment]
Update specification to prevent user exceptions from escaping

Update design document to note the possibility of back channels for exceptions

**Comment by jaredhoberock** [comment]
Update specification to return executor_future_t from sync_execute & bulk_sync_execute.

Update design doc to reflect that change.

**Comment by jaredhoberock** [comment]
Fixed by #189
